### PR TITLE
fix(streams): reroute boundary insertion through transform stream

### PR DESCRIPTION
**Pull request by AVaksman**
Fixes googleapis/google-api-nodejs-client/issues/1354

- [x] Tests and linter pass

Reroute `boundary` insertion to the the end of the stream via transform stream.
- To keep functionality separate, introduce a `BoundaryStream` class as extension of `transform` stream. 
- Pass `boundary` in the `_flush` method of the `BoundaryStream`.
- Eliminate the need to manually close `rStream`.

### CODE REVIEW

**Comment by AVaksman** [comment]
If it is OK to combine the functionality, then can add `_flush` to the `ProgressStream` class.

**Comment by AVaksman** [comment]
Will follow with Unit Test.

**Comment by JustinBeckwith** [comment]
👋 any story on tests here?  Would be awesome to see :)

File: src/apirequest.ts [hunk 1]
```diff
@@ -181,6 +181,7 @@ async function createAPIRequestAsync<T>(parameters: APIRequestParams) {
       const finale = `--${boundary}--`;
       const rStream = new stream.PassThrough();
       const pStream = new ProgressStream();
+      const bStream = new BoundaryStream(finale);
```
Comment (callmehiphop) [inline src/apirequest.ts]: `PassThrough` streams accept a `flush` function similarly to `Transform` streams. I think you could simplify the whole thing by just providing a `flush` function to `rStream`.

```js
const flush = callback => callback(null, `\r\n${finale}`);
const rStream = new stream.PassThrough({flush});
```

**Comment by AVaksman** [comment]
> I'm cool with this, but this is a very sensitive piece of code :) Please make sure before this lands that you:
> 
> * Add a unit test to ensure the desired behavior
> * `npm link` this change into `google-api-nodejs-client`, and run through the file multipart upload samples to make sure everything still works. Once that's done - happy to approve!

* Added unit test.
* Ran all samples in `google-api-nodejs-client` with these changes, everything still works.

**Pull request merged**